storage/zynq7000-flash/flashdrv.c

```diff
@@ -383,141 +342,439 @@ ssize_t flashdrv_read(addr_t offs, void *buff, size_t len, time_t timeout)
+		mutexLock(fdrv_common.regs[strg->dev->id].lock);
+		res = _flashdrv_chipErase(strg->dev->id);
+		mutexUnlock(fdrv_common.regs[strg->dev->id].lock);
+		return res;
+	}
+
+	mutexLock(fdrv_common.regs[strg->dev->id].lock);
+	while (len < size) {
+		res = _flashdrv_sectorErase(strg->dev->id, offs + strg->start + len);
+		if (res < 0) {
+			mutexUnlock(fdrv_common.regs[strg->dev->id].lock);
+			return res;
+		}
+
+		len += secSz;
+	}
+	mutexUnlock(fdrv_common.regs[strg->dev->id].lock);
+
+	return EOK;
+}
+
+
+static ssize_t flashdrv_mtdRead(struct _storage_t *strg, off_t offs, void *data, size_t len)
+{
+	int res;
+
+	mutexLock(fdrv_common.regs[strg->dev->id].lock);
+	res = _flashdrv_read(strg->dev->id, offs + strg->start, data, len);
+	mutexUnlock(fdrv_common.regs[strg->dev->id].lock);
+
+	return res;
+}
+
+
+static ssize_t flashdrv_mtdWrite(struct _storage_t *strg, off_t offs, const void *data, size_t len)
+{
+	ssize_t res = 0;
+	size_t tempSz = 0, chunkSz = 0;
+
+	mutexLock(fdrv_common.regs[strg->dev->id].lock);
+	chunkSz = offs % strg->dev->mtd->writeBuffsz;
+	if (chunkSz != 0) {
+		chunkSz = len < (strg->dev->mtd->writeBuffsz - chunkSz) ? len : (strg->dev->mtd->writeBuffsz - chunkSz);
+		res = _flashdrv_pageProgram(strg->dev->id, strg->start + offs, data, chunkSz);
+		if (res < 0) {
+			mutexUnlock(fdrv_common.regs[strg->dev->id].lock);
+			return res;
+		}
+
+		tempSz = res;
+	}
+
+	while (tempSz < len) {
+		chunkSz = ((len - tempSz) > strg->dev->mtd->writeBuffsz) ? strg->dev->mtd->writeBuffsz : (len - tempSz);
+		res = _flashdrv_pageProgram(strg->dev->id, strg->start + offs + tempSz, (const char *)data + tempSz, chunkSz);
+		if (res < 0) {
+			mutexUnlock(fdrv_common.regs[strg->dev->id].lock);
+			return res;
+		}
+
+		tempSz += res;
+	}
+	mutexUnlock(fdrv_common.regs[strg->dev->id].lock);
+
+	return tempSz;
+}
+
+
+const static storage_mtdops_t mtdOps = {
+	.erase = flashdrv_mtdErase,
+	.unPoint = NULL,
+	.point = NULL,
+	.read = flashdrv_mtdRead,
+	.write = flashdrv_mtdWrite,
+
+	.meta_read = NULL,
+	.meta_write = NULL,
+
+	.sync = NULL,
+	.lock = NULL,
+	.unLock = NULL,
+	.isLocked = NULL,
+
+	.block_isBad = NULL,
+	.block_isReserved = NULL,
+	.block_markBad = NULL,
+	.block_maxBadNb = NULL,
+
+	.suspend = NULL,
+	.resume = NULL,
+	.reboot = NULL,
+};
+
+
+/* Flash driver interface */
+
+int flashdrv_done(storage_t *strg)
+{
+	int i, res;
+
+	if (strg == NULL || strg->dev == NULL) {
+		return -EINVAL;
+	}
+
+	mutexLock(fdrv_common.regs[strg->dev->id].lock);
+	res = _flashdrv_sync(strg->dev->id, strg->start);
+
+	free(fdrv_common.regs[strg->dev->id].buff);
+	fdrv_common.regs[strg->dev->id].buff = NULL;
+	mutexUnlock(fdrv_common.regs[strg->dev->id].lock);
+
+	resourceDestroy(fdrv_common.regs[strg->dev->id].lock);
+	free(strg->dev->mtd);
+	free(strg->dev->blk);
+	free(strg->dev);
+
+	for (i = 0; i < fdrv_common.info.cfi.regsCount; ++i) {
+		if (fdrv_common.regs[i].buff != NULL) {
+			return res;
+		}
+	}
```

Comment:
Doesn't this need to be under lock too?

Reply:
`buff` is changed only in `flashdrv_done` so in this `for` there is no chance that other thread will change `buff`

Reply:
Ok, but above you agreed to protect buff under lock, but here is not needed? It's either needed in both cases or not needed at all. Can we have two concurrent flashdrv_done operations?

Reply:
Yeah, my mistake. We can assume that there will not be concurrent `flashdrv_done` operations. Restored previous code.